### ISSUE TITLE
place: mutex error on calls to `current-memory-use` with a custodian

### ISSUE THREAD

**Issue by Bogdanp** [racket/racket#4360]
This program crashes under Racket CS (HEAD) after a few iterations:

```racket
#lang racket/base

(require racket/place)
(provide start)

(define (start ch)
  (place-channel-get ch)
  (println "before custodian")
  (println "before gc")
  (collect-garbage) ;; makes it crash sooner
  (println "after gc")
  (define custodian (make-custodian))
  (current-memory-use custodian)
  (println "after custodian")
  (place-channel-put ch #t))

(module+ main
  (for ([_ (in-range 100)])
    (println "before place")
    (place-channel-put/get (dynamic-place "test.rkt" 'start) #t)
    (println "after place")))
```

with

```
"before place"
"before custodian"
"before gc"
"after gc"
"after custodian"
"after place"
"before place"
"before custodian"
"before gc"
mutex-release: thread does not own mutex 559348845
  context...:
   /var/folders/6s/8kt06x656dddy8z5y0jmf_fc0000gn/T/tmp.uBUsMl2U/test.rkt:6:0: start
```

The lock in question seems to be the `memory-limit-lock` in `thread/custodian.rkt`, but it's not obvious to me where it's being misused. Calling `current-memory-use` with a custodian seems to poison future garbage collections (presumably, it affects the next call to the increments callback, but that doesn't use the lock so I'm not sure what's going on).

Also, while I was looking at that module I noticed that the line

https://github.com/racket/racket/blob/569e2f4f47fee1043a19a8c2188d6981d09ea18a/racket/src/thread/custodian.rkt#L275

has a typo (`neeed/sent-wakeup` should be `needed/sent-wakeup`, though no code expects that specific value so it's not causing any problems).

**Comment by mflatt** [comment]
Thanks for the report!

The problem is not really with a lock, but with a misuse of an accessor in unsafe mode that cascades to memory corruption. Specifically, the problem is that the argument to `call-with-size-increments` receives a `#!bwp` in the `custs` list, so the `custodian-memory-use` accessor fails.

For anyone interested, some hints on debugging a problem like this: I got one step closer by using `zuo . run UNSAFE_COMP=no` within `racket/src/cs`. That compiles in safe mode, so that `custodian-memory-use` will detect the misapplication. Safe-mode compilation changes the symptom to `$collect-rendezvous: cannot return to the collect-request-handler`, which is still not the real problem, but turns out to be a sign that there was an exception in the garbage-collection callback. I'll add a hint comment in `collect/report` on catching and printing errors during the collection callback to get more information, and that printout is how I got `Exception in custodian-memory-use: #!bwp is not of type #<record type custodian>`.
